src/Components/Web/src/HtmlRendering/StaticHtmlRenderer.HtmlWriting.cs

```diff
@@ -173,6 +178,29 @@ private int RenderElement(int componentId, TextWriter output, ArrayRange<RenderT
         }
     }
 
+    private int RenderScriptElementChildren(int componentId, TextWriter output, ArrayRange<RenderTreeFrame> frames, int position, int maxElements)
+    {
+        // Script tags are special. Unlike in XHTML, the contents of script elements in HTML
+        // are parsed as plain text, so HTML entities are not recognized and are treated as
+        // literal text. Code like console.log(&quot;Hello&quot;) would be a syntax error.
+        // HTML does not make use of <![CDATA[ ... ]]> either. So inside <script>...</script>,
+        // we must disable HTML encoding. This is hardly a XSS issue since if an application
+        // is rendering user content inside <script>, they already have an XSS problem unless
+        // they have carefully escaped the user content.
+        // https://stackoverflow.com/a/14781466
+        // https://html.spec.whatwg.org/multipage/scripting.html#restrictions-for-contents-of-script-elements
```

Comment:
So after more investigation, if our goal is to be consistent with `.cshtml`, then it's best to understand this as a Razor compiler limitation: https://github.com/dotnet/razor/issues/9204. We would need a compiler change to be able to know when it's safe to emit some unencoded content into `<script>` and when it is not.

If that is the case, then I think that means we have to not support it in the short term. So my proposed plan is:

 * Keep the parts of this PR that make it possible to add `<script>` tags at all
 * Remove the parts that deal with HTML encoding, so it will only work for `<script src='someUrl'>` and `<script>@(new MarkupString("myscript"))</script>`
 * In the future, maybe https://github.com/dotnet/razor/issues/9204 will get fixed and then it would automatically just start working for `<script>code</script>` as well.

Reply:
OK, with a bit of extra insight from https://github.com/dotnet/razor/issues/9204, I think we can do a bit better than the above, and can actually make everything work with the correct encodings [*]

@javiercn's original suggestion to use `JavaScriptEncoder` was correct. It doesn't break cases where the developer's compile-time constant code gets compiled as `AddMarkupContent`, since that's emitted verbatim. All it does is force any dynamic expressions to be JavaScript-string encoded, which is correct and equivalent to the HTML-encoding that we enforce for dynamic expressions outside `<script>` contexts.

So altogether I think this PR is ready now.

* [*] Until https://github.com/dotnet/razor/issues/9204 is fixed, single-line `<script>` blocks will have too much encoding (even compile-time constant code will get JavaScript encoded), which is safe but prevents the code from executing. Developers can work around this by adding at least one linebreak. The issue will go away when that Razor compiler issue is fixed.